### PR TITLE
Add a type-erased Vamana index (which only supports creation)

### PR DESCRIPTION
**Pull request by jparismorgan**
### What
Here we add a type-erased Vamana index which only supports creation. In a follow-up I will add support for querying and ingestion. I do this because this PR already has a good amount of interesting changes, so I want to keep it as simple as possible. Changes included:
* Add `vamana_index.py` which in `create()` will create an empty Vamana index and write it to the specific URI
* Fix crashes in `write_matrix()` and `write_vector()` that would happen during the `index/vamana_index.h@write_index()` call when we try to write an empty vector.
* Update the Vamana index (i.e. `vamana_group.h`) to use `parts_array_name` to store feature vectors instead of a custom `feature_vectors_array_name`. This matches what FLAT does and will make the Python code simpler.
  * We also delete some tests from `unit_vamana_group.cc` because of this - these fail because the previously created indexes use `feature_vectors_array_name`.
* Update the Vamana index (i.e. `index/vamana_index.h`) to store IDs in `ids_array_name`, and update to `ColMajorMatrixWithIds<feature_type, id_type> feature_vectors_;` along with that so we can store the IDs in the index.

### Testing
* Existing unit tests pass.
* New unit tests pass.

### CODE REVIEW

**Comment by jparismorgan** [comment]
Opening up this PR for review before `Python API CI` is passing b/c that is a separate issue which is being discussed here: https://tiledb.slack.com/archives/C06H3JS58AD/p1710935849456319?thread_ts=1710777709.344369&cid=C06H3JS58AD

I will wait until it passes to merge this, but we can still get this reviewed in the meantime.

File: apis/python/src/tiledb/vector_search/vamana_index.py
Comment (NikolaosPapailiou) [inline apis/python/src/tiledb/vector_search/vamana_index.py:29]: nit: remove `memory_budget`

File: apis/python/src/tiledb/vector_search/vamana_index.py
Comment (NikolaosPapailiou) [inline apis/python/src/tiledb/vector_search/vamana_index.py:45]: I don't think that `index_version` is needed for this new type of index. It is used only for backwards compatibility for the existing indexes.

File: apis/python/src/tiledb/vector_search/vamana_index.py
Comment (NikolaosPapailiou) [inline apis/python/src/tiledb/vector_search/vamana_index.py:72]: Lets add the extra query args only when we have an actual distributed query implementation.

File: apis/python/src/tiledb/vector_search/vamana_index.py
Comment (NikolaosPapailiou) [inline apis/python/src/tiledb/vector_search/vamana_index.py:151]: I think we should try to simplify this code by providing one method in the C++ API that creates an empty index.

File: apis/python/src/tiledb/vector_search/vamana_index.py
Comment (jparismorgan) [reply apis/python/src/tiledb/vector_search/vamana_index.py]: Thanks, done.

File: apis/python/src/tiledb/vector_search/vamana_index.py
Comment (jparismorgan) [reply apis/python/src/tiledb/vector_search/vamana_index.py]: Thanks, done.

File: apis/python/src/tiledb/vector_search/vamana_index.py
Comment (jparismorgan) [reply apis/python/src/tiledb/vector_search/vamana_index.py:64]: Thanks, done.

File: apis/python/src/tiledb/vector_search/vamana_index.py
Comment (jparismorgan) [reply apis/python/src/tiledb/vector_search/vamana_index.py:116]: Agree, though for now I'd prefer to leave as-is just for now - I don't yet have everything working end to end and it's faster to make changes in Python than in native code. Added a TODO and will update once it's fully working, if okay with you?

**Review by NikolaosPapailiou — APPROVED** [review]

**Pull request merged**
